src/Evolution/Systems/GrMhd/ValenciaDivClean/ComovingMagneticFieldMagnitude.cpp

```diff
@@ -0,0 +1,62 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/ComovingMagneticFieldMagnitude.hpp"
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
+#include "PointwiseFunctions/GeneralRelativity/IndexManipulation.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace grmhd::ValenciaDivClean::Tags {
+
+void ComovingMagneticFieldMagnitudeCompute::function(
+    const gsl::not_null<Scalar<DataVector>*> comoving_magnetic_field_magnitude,
+    const tnsr::I<DataVector, 3>& magnetic_field,
+    const tnsr::I<DataVector, 3>& spatial_velocity,
+    const Scalar<DataVector>& lorentz_factor,
+    const tnsr::ii<DataVector, 3, Frame::Inertial>& spatial_metric) {
+  Variables<tmpl::list<hydro::Tags::SpatialVelocityOneForm<DataVector, 3>,
+                       hydro::Tags::MagneticFieldDotSpatialVelocity<DataVector>,
+                       hydro::Tags::MagneticFieldOneForm<DataVector, 3>,
+                       hydro::Tags::MagneticFieldSquared<DataVector>>>
+      temp_tensors{get(lorentz_factor).size()};
+
+  const auto& spatial_velocity_one_form =
+      get<hydro::Tags::SpatialVelocityOneForm<DataVector, 3>>(temp_tensors);
+  raise_or_lower_index(
+      make_not_null(&get<hydro::Tags::SpatialVelocityOneForm<DataVector, 3>>(
+          temp_tensors)),
+      spatial_velocity, spatial_metric);
+
+  const auto& magnetic_field_dot_spatial_velocity =
+      get<hydro::Tags::MagneticFieldDotSpatialVelocity<DataVector>>(
+          temp_tensors);
+  dot_product(
```

Comment:
why not use the `dot_product` that takes the metric directly instead of first explicitly lowering the spatial velocity?

Reply:
My bad! :b